Turn on typechecking for authzCourseOrInstance

nwalters512 opened the pull request:
The main change that had to be made here was figuring out how to make TypeScript happy with attaching "info" to errors. I opted to introduce a new `AugmentedError` class that I hope will eventually take the place of all our `makeWith...` error utilities.

nwalters512 commented on packages/error/src/index.ts:
```diff
@@ -53,3 +53,23 @@ export function newMessage(err: any, newMsg: string): ErrorWithData {
+  cause?: unknown;
+}
+
+export class AugmentedError extends Error {
```
I borrowed the "augment" term from #7968, but I'm open to other ideas.

mwest1066 replied:
Works for me!

mwest1066 commented on apps/prairielearn/src/middlewares/authzCourseOrInstance.js:
```diff
@@ -164,14 +165,15 @@ module.exports = asyncHandler(async (req, res, next) => {
-      </p>
-    `.toString();
-    throw err;
+    throw new error.AugmentedError('Access denied', {
```
Not that I mind, but I think we'd normally just import `AugmentedError` directly?

nwalters512 replied:
Yep, this was me + Copilot being lazy 😂 I'll update this to do the right thing here.

nwalters512 replied:
Done in ea136d0.

mwest1066 commented on packages/error/src/index.ts:
```diff
@@ -53,3 +53,23 @@ export function newMessage(err: any, newMsg: string): ErrorWithData {
+
+export interface AugmentedErrorOptions {
+  status?: number;
+  info?: string;
```
Would it be better to have this as `HTMLSafeString | string` so that we can template it into HTML without having to worry about safety? That is, at the moment we presumably need to template it into HTML without further escaping, which is fine if we've already escaped it but not fine if we started with an unsafe string. I'm not sure that's the world's best explanation but I hope you see what I mean 😅

nwalters512 replied:
I do see what you mean, and if our error page lived in `@prairielearn/html`-land, I'd agree with you. Unfortunately, it's still written in EJS. I like the idea; let's save it for the future?

mwest1066 replied:
Ok, I see we are using unescaped `<%- error.info %>` on the error page. This feels pretty dangerous but I guess we haven't managed to shoot ourselves in the foot yet with this 😅

nwalters512 replied:
Well, we did, but I actually changed that here: https://github.com/PrairieLearn/PrairieLearn/pull/8010

Here's a thought: make `status` _only_ accept `HTMLSafeString`, but immediately `toString()` it? Alternatively, I guess we could just unconditionally `toString()` it in the error page? I think either way, we could do better without having to refactor the error page to not use EJS. Let me play around with this.

nwalters512 replied:
I made this change in f749ba2, it works great!

mwest1066 commented:
If we are going to use this instead of `augmentError()` from #7968, could we add defaults:
* `status` defaults to 500
* `message` defaults to `cause.message` if we have `cause` and don't have `message`

Given these two enhancements I think we should then delete `augmentError()`.

nwalters512 commented:
@mwest1066 I introduced a default status in ea136d0; I also added an `HttpStatusError` as discussed on Slack. I did not make any changes to make `message` optional and pull it from the cause if specified, as I think we should always be attaching our own meaningful messages to errors. Let me know if you disagree with that.

mwest1066 commented:
> @mwest1066 I introduced a default status in [ea136d0](https://github.com/PrairieLearn/PrairieLearn/commit/ea136d0229f793026cb7a26e40339ef423abc38e); I also added an `HttpStatusError` as discussed on Slack. I did not make any changes to make `message` optional and pull it from the cause if specified, as I think we should always be attaching our own meaningful messages to errors. Let me know if you disagree with that.

Yep, I agree with your reasoning about the message.
